File: apps/web/src/lib/api/citations.ts
import { request } from './client';
import type { Author, CitationItem, CitationStyle, AttributionScope, ExtractionStatus } from '@openresearch/citations';
import type { PaperDTO } from './papers';

export interface CitationDTO {
  id: string;
  document_id: string;
  paper_id: string;
  position: number;
  citation_style: CitationStyle | string;
  attribution_scope: AttributionScope | string;
  page_number?: number;
  relevant_passage?: string;
  created_at: string;
  paper?: PaperDTO;
}

export interface ResolvedIdentifierDTO {
  identifier: string;
  id_type: string;
  title?: string;
  authors: Author[];
  year?: number;
  abstract?: string;
  doi?: string;
  arxiv_id?: string;
  pmid?: string;
  journal?: string;
  publisher?: string;
  volume?: string;
  issue?: string;
  pages?: string;
  url?: string;
  bibtex?: string;
  extraction_status: ExtractionStatus;
}

export interface RankedCitationDTO {
  paper_id: string;
  title: string;
  authors: Author[];
  year?: number;
  score: number;
  extraction_status: ExtractionStatus;
}

export const citationsApi = {
  list: (documentId: string) => request<CitationDTO[]>(`/documents/${documentId}/citations`),
  create: (
    documentId: string,
    data: {
      paper_id: string;
      position?: number;
      citation_style?: CitationStyle | string;
      attribution_scope?: AttributionScope | string;
      page_number?: number;
      relevant_passage?: string;
    }
  ) =>
    request<CitationDTO>(`/documents/${documentId}/citations`, {
      method: 'POST',
      body: JSON.stringify({ document_id: documentId, ...data }),
    }),
  delete: (documentId: string, citationId: string) =>
    request<void>(`/documents/${documentId}/citations/${citationId}`, {
      method: 'DELETE',
    }),
  resolveIdentifier: (identifier: string, idType: string = 'auto') =>
    request<ResolvedIdentifierDTO>('/citations/resolve-identifier', {
      method: 'POST',
      body: JSON.stringify({ identifier, id_type: idType }),
    }),
  addByIdentifier: (projectId: string, identifier: string, idType: string = 'auto') =>
    request<PaperDTO>(`/projects/${projectId}/papers/add-by-identifier`, {
      method: 'POST',
      body: JSON.stringify({ project_id: projectId, identifier, id_type: idType }),
    }),
  importBibtex: (projectId: string, bibtexContent: string) =>
    request<{ total_imported: number; papers: PaperDTO[] }>(`/projects/${projectId}/papers/import-bibtex`, {
      method: 'POST',
      body: JSON.stringify({ project_id: projectId, bibtex_content: bibtexContent }),
    }),
  exportProjectBibtex: (projectId: string) =>
    request<{ bibtex_content: string; total_entries: number }>(`/projects/${projectId}/export/bibtex`),
  exportDocumentBibtex: (documentId: string) =>
    request<{ bibtex_content: string; total_entries: number }>(`/documents/${documentId}/export/bibtex`),
};

File: apps/api/app/schemas/citations.py
"""Citation & BibTeX schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from app.schemas.papers import PaperResponse


class CitationCreate(BaseModel):
    document_id: str
    paper_id: str
    position: int = 0
    citation_style: str = "apa"
    attribution_scope: str = "sentence"
    page_number: int | None = None
    relevant_passage: str | None = None


class CitationResponse(BaseModel):
    id: str
    document_id: str
    paper_id: str
    position: int
    citation_style: str
    attribution_scope: str
    page_number: int | None = None
    relevant_passage: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CitationDetailResponse(BaseModel):
    id: str
    document_id: str
    paper_id: str
    position: int
    citation_style: str
    attribution_scope: str
    page_number: int | None = None
    relevant_passage: str | None = None
    paper_title: str | None = None
    authors: list[dict[str, Any]] | None = None
    year: int | None = None
    doi: str | None = None
    extraction_status: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class IdentifierResolveRequest(BaseModel):
    identifier: str
    id_type: str | None = "auto"  # 'auto' | 'doi' | 'arxiv' | 'pmid'


class IdentifierResolveResponse(BaseModel):
    identifier: str
    id_type: str  # 'doi' | 'arxiv' | 'pmid'
    title: str | None = None
    authors: list[dict[str, Any]] = []
    year: int | None = None
    abstract: str | None = None
    doi: str | None = None
    arxiv_id: str | None = None
    pmid: str | None = None
    journal: str | None = None
    publisher: str | None = None
    volume: str | None = None
    issue: str | None = None
    pages: str | None = None
    url: str | None = None
    bibtex: str | None = None
    extraction_status: str = "ok"


class AddByIdentifierRequest(BaseModel):
    project_id: str
    identifier: str
    id_type: str | None = "auto"


class BibtexImportRequest(BaseModel):
    project_id: str
    bibtex_content: str


class BibtexImportResponse(BaseModel):
    total_imported: int
    papers: list[PaperResponse]


class BibtexExportResponse(BaseModel):
    bibtex_content: str
    total_entries: int


class ContextRankingRequest(BaseModel):
    document_id: str
    paragraph_text: str
    query: str | None = None
    limit: int | None = 10


class ContextRankedPaper(BaseModel):
    paper_id: str
    title: str
    authors: list[dict[str, Any]]
    year: int | None = None
    score: float
    extraction_status: str


class ContextRankingResponse(BaseModel):
    results: list[ContextRankedPaper]
